File: lib/src/api/dashboard/dashboard.controller.ts
import {
  Body,
  Controller,
  Delete,
  Get,
  Param,
  Patch,
  Post,
  Query,
  UseGuards
} from "@nestjs/common";
import { DashboardQueryService } from "./dashboard-query/dashboard-query.service";
import { DashboardService } from "./dashboard.service";
import { Dashboard } from "../../db/entities/dashboard.entity";
import { ApiResponse } from "../../common/types/dto/response.dto";
import { DashboardPanel } from "../../db/entities/dashboard-panel.entity";
import {
  DashboardDto,
  DashboardPanelDto,
  LayoutChangeDto
} from "../../common/types/dto/dashboard.dto";
import { ApiTags } from "@nestjs/swagger";
import { AuthGuard } from "../../common/decorators/auth-guard.decorator";

@ApiTags("dashboard")
@Controller("dashboard")
@UseGuards(new AuthGuard())
export class DashboardController {
  constructor(
    private readonly dashboardQueryService: DashboardQueryService,
    private readonly dashboardService: DashboardService
  ) {}

  @Get("/:id")
  private async getDashboard(@Param("id") id: string): Promise<ApiResponse<Dashboard>> {
    return await this.dashboardQueryService.getDashboard(id);
  }

  // TODO: do i need this endpoint?
  @Get("/:id/panels")
  private async getDashboardPanels(@Param("id") id: string): Promise<ApiResponse<Dashboard>> {
    return await this.dashboardQueryService.getDashboardPanels(id);
  }

  @Get("/panel/:id")
  private async getDashboardPanel(@Param("id") id: string): Promise<ApiResponse<DashboardPanel>> {
    return await this.dashboardQueryService.getDashboardPanel(id);
  }

  @Get("/project/:id")
  private async getProjectDashboards(
    @Param("id") id: string
  ): Promise<ApiResponse<DashboardPanel>> {
    return await this.dashboardQueryService.getProjectDashboards(id);
  }

  //  //  //  //

  @Post()
  private async createDashboard(@Body() body: DashboardDto): Promise<ApiResponse<string>> {
    return await this.dashboardService.createDashboard(body);
  }

  @Post("/panel")
  private async createDashboardPanel(
    @Body() body: DashboardPanelDto
  ): Promise<ApiResponse<string>> {
    return await this.dashboardService.createPanel(body);
  }

  //  //  //  //

  @Patch()
  private async updateDashboard(@Body() body: DashboardDto): Promise<ApiResponse<string>> {
    return await this.dashboardService.updateDashboard(body);
  }

  @Patch("/panel")
  private async updatePanel(@Body() body: DashboardPanelDto): Promise<ApiResponse<string>> {
    return await this.dashboardService.updatePanel(body);
  }

  @Patch("/layout")
  private async updateDashboardLayout(
    @Body() body: LayoutChangeDto
  ): Promise<ApiResponse<string>> {
    return await this.dashboardService.updateDashboardLayout(body);
  }

  //  //  //  //

  @Delete()
  private async removeDashboard(
    @Query("dashboardId") dashboardId: string,
    @Query("projectId") projectId: string
  ): Promise<ApiResponse<string>> {
    return await this.dashboardService.removeDashboard(dashboardId, projectId);
  }

  @Delete("/panel/:id")
  private async removePanel(@Param("id") id: string): Promise<ApiResponse<string>> {
    return await this.dashboardService.removePanel(id);
  }
}

File: server/traceo-api/src/main/java/org/traceo/api/controllers/DashboardController.java
package org.traceo.api.controllers;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.traceo.api.models.response.CreateResponse;
import org.traceo.api.services.commands.DashboardPanelService;
import org.traceo.api.services.commands.DashboardService;
import org.traceo.api.services.queries.DashboardQueryService;
import org.traceo.common.transport.dto.api.DashboardDto;
import org.traceo.common.transport.dto.api.DashboardLayoutDto;
import org.traceo.common.transport.dto.api.DashboardPanelDto;
import org.traceo.common.transport.response.ApiResponse;

import java.util.List;

@RestController
@RequestMapping("/api/dashboard")
public class DashboardController {
    private final DashboardService dashboardService;
    private final DashboardPanelService dashboardPanelService;
    private final DashboardQueryService dashboardQueryService;

    public DashboardController(DashboardService dashboardService, DashboardPanelService dashboardPanelService, DashboardQueryService dashboardQueryService) {
        this.dashboardService = dashboardService;
        this.dashboardPanelService = dashboardPanelService;
        this.dashboardQueryService = dashboardQueryService;
    }

    @GetMapping("/{id}")
    private ResponseEntity<ApiResponse> getDashboard(@PathVariable String id) {
        DashboardDto response = dashboardQueryService.getDashboard(id);
        return new ResponseEntity<>(ApiResponse.ofSuccess(response), HttpStatus.OK);
    }

    @GetMapping("/{id}/panels")
    private ResponseEntity<ApiResponse> getDashboardPanels(@PathVariable String id) {
        List<DashboardPanelDto> response = dashboardQueryService.getDashboardPanels(id);
        return new ResponseEntity<>(ApiResponse.ofSuccess(response), HttpStatus.OK);
    }

    @GetMapping("/panel/{id}")
    private ResponseEntity<ApiResponse> getDashboardPanel(@PathVariable String id) {
        DashboardPanelDto response = dashboardQueryService.getDashboardPanel(id);
        return new ResponseEntity<>(ApiResponse.ofSuccess(response), HttpStatus.OK);
    }

    @PostMapping
    private ResponseEntity<ApiResponse> createDashboard(@RequestBody DashboardDto dto) {
        String id = dashboardService.create(dto);
        return new ResponseEntity<>(ApiResponse.ofSuccess(new CreateResponse((id))), HttpStatus.CREATED);
    }

    @PostMapping("/panel")
    private ResponseEntity<ApiResponse> createDashboardPanel(@RequestBody DashboardPanelDto dto) {
        String id = dashboardPanelService.create(dto);
        return new ResponseEntity<>(ApiResponse.ofSuccess(new CreateResponse((id))), HttpStatus.CREATED);
    }

    @PatchMapping()
    private ResponseEntity<ApiResponse> updateDashboard(@RequestBody DashboardDto dto) {
        dashboardService.update(dto);
        return new ResponseEntity<>(ApiResponse.ofSuccess(), HttpStatus.NO_CONTENT);
    }

    @PatchMapping("/panel")
    private ResponseEntity<ApiResponse> updateDashboardPanel(@RequestBody DashboardPanelDto dto) {
        dashboardPanelService.update(dto);
        return new ResponseEntity<>(ApiResponse.ofSuccess(), HttpStatus.NO_CONTENT);
    }

    @PatchMapping("/layout")
    private ResponseEntity<ApiResponse> updateDashboardLayout(@RequestBody DashboardLayoutDto dto) {
        dashboardService.updateLayout(dto);
        return new ResponseEntity<>(ApiResponse.ofSuccess(), HttpStatus.NO_CONTENT);
    }

    @DeleteMapping()
    private ResponseEntity<ApiResponse> removeDashboard(@RequestParam String dashboardId, @RequestParam String projectId) {
        dashboardService.remove(dashboardId);
        return new ResponseEntity<>(ApiResponse.ofSuccess(), HttpStatus.NO_CONTENT);
    }

    @DeleteMapping("/panel/{id}")
    private ResponseEntity<ApiResponse> removePanel(@PathVariable String id) {
        dashboardPanelService.remove(id);
        return new ResponseEntity<>(ApiResponse.ofSuccess(), HttpStatus.NO_CONTENT);
    }
}
